pkg/operator/operator.go

```diff
@@ -241,76 +273,237 @@ func (k *Kontroller) withLeaderElection() error {
+	glog.V(4).Info("Marking rebooted nodes with after-reboot label")
+	err = k.markAfterReboot()
+	if err != nil {
+		glog.Errorf("Failed to update recently rebooted nodes: %v", err)
+		return
+	}
 
-	justRebootedNodes := k8sutil.FilterNodesByAnnotation(nodelist.Items, justRebootedSelector)
+	// find nodes with the before-reboot=true label and check if all provided
+	// annotations are set. if all annotations are set to true then remove the
+	// before-reboot=true label and set reboot=ok=true, telling the agent it's
+	// time to reboot.
+	glog.V(4).Info("Checking for existence of configured before-reboot annotations")
+	err = k.checkBeforeReboot()
+	if err != nil {
+		glog.Errorf("Failed to check before reboot: %v", err)
+		return
+	}
 
-	if len(justRebootedNodes) > 0 {
-		glog.Infof("Found %d rebooted nodes, setting annotation %q to false", len(justRebootedNodes), constants.AnnotationOkToReboot)
+	// take some number of the rebootable nodes. remove any old configured
+	// annotations and add the before-reboot=true label.
+	glog.V(4).Info("Marking rebootable nodes with before-reboot label")
+	err = k.markBeforeReboot()
+	if err != nil {
+		glog.Errorf("Failed to update rebootable nodes: %v", err)
+		return
 	}
+}
 
-	for _, n := range justRebootedNodes {
-		glog.Infof("Setting 'ok-to-reboot=false' for %q", n.Name)
-		if err := k8sutil.SetNodeAnnotations(k.nc, n.Name, map[string]string{
-			constants.AnnotationOkToReboot: constants.False,
-		}); err != nil {
-			glog.Infof("Failed setting annotation %q on node %q to false: %v", constants.AnnotationOkToReboot, n.Name, err)
+// checkBeforeReboot gets all nodes with the before-reboot=true label and checks
+// if all of the configured before-reboot annotations are set to true. If they
+// are, it deletes the before-reboot=true label and sets reboot-ok=true to tell
+// the agent that it is ready to start the actual reboot process.
+// If it goes to set reboot-ok=true and finds that the node no longer wants a
+// reboot, then it just deletes the before-reboot=true label.
+// If there is an error getting the list of nodes or updating any of them, an
+// error is immediately returned.
+func (k *Kontroller) checkBeforeReboot() error {
+	nodelist, err := k.nc.List(v1meta.ListOptions{})
+	if err != nil {
+		return fmt.Errorf("Failed listing nodes: %v", err)
+	}
+
+	preRebootNodes := k8sutil.FilterNodesByRequirement(nodelist.Items, beforeRebootReq)
+
+	for _, n := range preRebootNodes {
+		if hasAllAnnotations(n, k.beforeRebootAnnotations) {
+			glog.V(4).Infof("Deleting label %q for %q", constants.LabelBeforeReboot, n.Name)
+			glog.V(4).Infof("Setting annotation %q to true for %q", constants.AnnotationOkToReboot, n.Name)
+			err = k8sutil.UpdateNodeRetry(k.nc, n.Name, func(node *v1api.Node) {
+				delete(node.Labels, constants.LabelBeforeReboot)
+
+				// make sure that the node still wants to reboot
+				if !wantsRebootSelector.Matches(fields.Set(node.Annotations)) {
```

Comment:
As far as cleaning up before-reboot and after-reboot annotations, it's currently doing it right before it applies the respective label to the node, to ensure an entirely clean environment, and that there aren't any old annotations lying around from the last run. It seems to me like if we only clean them up when we remove the label after completion, there is a possibility that a stray annotation might make it into the mix. Maybe I'm being paranoid.